universe/Condition.h

```diff
@@ -88,9 +82,10 @@ struct FO_COMMON_API Condition {
     { return 0; }
 
 protected:
-    Invariance m_root_candidate_invariant = UNKNOWN_INVARIANCE;
-    Invariance m_target_invariant = UNKNOWN_INVARIANCE;
-    Invariance m_source_invariant = UNKNOWN_INVARIANCE;
+    bool m_root_candidate_invariant = false;
+    bool m_local_candidate_invariant = false;
```

Comment:
What's the purpose of `m_local_candidate_invariant` ? I think it will always be false...

Reply:
Some `Condition` subclasses calculate the local candidate invariance in their `Condition::Eval` method. Calculating the local candidate invariance should also be done in the constructor, which requires the serialization of the invariances.  As I am already introducing serialization for the other invariances I didn't want to break the serialization twice in short succession.

Reply:
You mean the `simple_eval_safe` tests based on subconditions and valuerefs?

That's not the same as a the whole condition being local candidate invariant, and probably doesn't need to be exposed with a getter. It also probably shouldn't be part of the base class, as its an implementation detail of some of the derived classes, but not needed for others. It might also be replaced on a case-by-case basis with something more complicated... eg. there could be conditions with multiple subconditions or valuerefs, some of which are local candidate invariant and some which aren't, and the `Eval` function could re-evaluate the variant ones but not the invariant ones, rather than summarizing all of them as invariant or not.

Reply:
> You mean the simple_eval_safe tests based on subconditions and valuerefs?

Yes, I was talking about those.

The problem I have is that the implementation is overcomplicated.  You cannot understand 10.000 + 5000 lines of code.

Reply:
It is a huge complicated system. But it's not as bad as it might seem, as there is a lot of just-slightly-different functions and code between the various classes.

Reply:
Removed `Condition::Condition::m_local_candidate_invariant` attribute again.

Reply:
> If you want to get into slightly more in-depth refactoring, it might make sense to combine some conditions that have similar interfaces into a single multi-purpose condition, that uses an enum to control how it functions.

Not now, this PR was mind-numbing enough for now.

> For example, I think RootCandidate, Target, Capital, Monster, Armed, Stationary, Aggressive, CanColonzie, CanProduceShips, All, None, and Source don't have any internal state and don't take any parameters when constructed, so don't really need to be separate classes...

> Sort like how ValueRef classes work, the Eval and Match functions would check the enum value to decide how to evaluate their input, and might get fairly big. But the Description or Dump would just refer to some members storing what strings to use, which would set during construction. Some external code that relies on casting pointers would need to further inspect the condition, like in the objects list window, as well.

Sounds good and matches what I had in mind but I would prefer to use member function pointers or functors for this instead of enums.  Trying to mimicking the interface of boost::python function bindings and removing most of the Condition/ValueRef implementations.

Also there are various implementations like `Building`, `Planet{Type,Size}` or Species, which are pretty much identical, comparing an entity property against a list of values.  Things like that are certainly not neccessary and can be replaced within the parser by binding the matching member function/functor to a more generalized `InList` Condition.